ui/Modal/Org/ConfigureEns/RegisterName.svelte

```diff
@@ -0,0 +1,196 @@
+<!--
+ Copyright © 2021 The Radicle Upstream Contributors
+
+ This file is part of radicle-upstream, distributed under the GPLv3
+ with Radicle Linking Exception. For full terms see the included
+ LICENSE file.
+-->
+<script lang="typescript" context="module">
+  export interface Result {
+    registration: ensResolver.Registration | null;
+    name: string;
+  }
+</script>
+
+<script lang="typescript">
+  import * as ethers from "ethers";
+
+  import { unreachable } from "ui/src/unreachable";
+  import * as svelteStore from "ui/src/svelteStore";
+  import * as wallet from "ui/src/wallet";
+
+  import { Modal, TextInput } from "ui/DesignSystem";
+
+  import ConfirmRegistration from "./ConfirmRegistration.svelte";
+  import ButtonRow from "./ButtonRow.svelte";
+
+  import * as ensRegistrar from "ui/src/org/ensRegistrar";
+  import * as ensResolver from "ui/src/org/ensResolver";
+  import * as error from "ui/src/error";
+  import * as validation from "ui/src/validation";
+
+  export let registrationDone: (result: Result) => void;
+  export let currentName: string | undefined;
+  export let fee: ethers.BigNumber;
+
+  let nameInputValue: string = currentName || "";
+
+  type State =
+    | {
+        type: "validateAndCommit";
+      }
+    | {
+        type: "register";
+        commitmentSalt: Uint8Array;
+        commitmentBlock: number;
+        minAge: number;
+      };
+
+  let state: State = { type: "validateAndCommit" };
+
+  let validationStatus: validation.ValidationState = {
+    status: validation.ValidationStatus.NotStarted,
+  };
+
+  let timeoutHandle: number;
+  let userInputStarted: boolean = nameInputValue !== "";
+  let registration: ensResolver.Registration | null;
+
+  $: validateName(nameInputValue);
+
+  function validateName(nameInputValue: string | undefined): void {
+    if (!userInputStarted) {
+      userInputStarted = true;
+      return;
+    }
+
+    if (!nameInputValue) {
+      validationStatus = {
+        status: validation.ValidationStatus.Error,
+        message: "You need to enter a name.",
+      };
+    } else {
+      validationStatus = {
+        status: validation.ValidationStatus.Loading,
+      };
+      if (timeoutHandle) {
+        clearTimeout(timeoutHandle);
+      }
+      timeoutHandle = window.setTimeout(() => {
+        validateFormAndQueueNextAction();
+      }, 1000);
```

Comment:
Although we cancel the timeout there is still a potential race condition here: Assume the user types "x", waits more then a second, then types "y". This will call `validateFormAndQueueNextAction` twice—first, just before typing "y" and then a second later. Unfortunately it’s not guaranteed that the first async call finishes before the second call finishes. `ensRegistrar.isAvailable()` may take more then a second the first time and be very quick the second time around. This results in the UI being updated with the result of the first validation call for "x".

I suggest we use `lodash.debounce()` together with `MutexWorker` to guard against this. (`MutexExecutor` would work, too.) Happy to pair on this if you need support.

Reply:
5150e5d595096eec92d61958f4922514b63a4da3